### PR TITLE
Address MQTT MISRA 15

### PR DESCRIPTION
**Pull request by muneebahmed10**
*Issue #, if available:*

*Description of changes:*
Properly annotate the exception for 14.3, and refactor the multiple breaks/returns for 15.4 and 15.5

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/common/include/iot_linear_containers.h [hunk 1]
```diff
@@ -623,7 +623,8 @@ static inline IotLink_t * IotListDouble_FindFirstMatch( const IotListDouble_t *
 {
     /* The const must be cast away to match this function's return value. Nevertheless,
      * this function will respect the const-ness of pStartPoint. */
-    IotLink_t * pCurrent = ( IotLink_t * ) pStartPoint;
+    IotLink_t * pCurrent = ( IotLink_t * ) pStartPoint, * pMatchedLink = NULL;
+    bool shouldExit = false;
```
Comment (gordonwang0) [inline libraries/standard/common/include/iot_linear_containers.h]: Minor: rename to `matchFound`.

File: libraries/standard/mqtt/src/iot_mqtt_subscription.c [hunk 2]
```diff
@@ -230,7 +230,7 @@ static bool _topicFilterMatch( const char * pTopicName,
                                const char * pTopicFilter,
                                uint16_t topicFilterLength )
 {
-    bool status = false;
+    bool status = false, shouldBreak = false;
```
Comment (gordonwang0) [inline libraries/standard/mqtt/src/iot_mqtt_subscription.c]: Minor: rename to `matchFound`.

File: libraries/standard/mqtt/src/iot_mqtt_api.c [hunk 3]
```diff
@@ -984,6 +984,7 @@ IotMqttError_t IotMqtt_Init( void )
          * condition is not an invariant, and the MISRA 14.3 violation is
          * a false positive. */
         /* coverity[misra_c_2012_rule_14_3_violation] */
+        /* coverity[const] */
```
Comment (abhidixi11) [inline libraries/standard/mqtt/src/iot_mqtt_api.c]: Why const exception is needed ? Same reason as 14.3 ?

File: libraries/standard/common/include/iot_linear_containers.h (hunk 1 above)
Comment (muneebahmed10) [reply libraries/standard/common/include/iot_linear_containers.h]: Done.

File: libraries/standard/mqtt/src/iot_mqtt_subscription.c (hunk 2 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/src/iot_mqtt_subscription.c]: Done

File: libraries/standard/mqtt/src/iot_mqtt_api.c (hunk 3 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/src/iot_mqtt_api.c]: Yes, there were two events for that violation:
```
(1) Event const: | At condition "status == IOT_MQTT_SUCCESS", the value of "status" must be equal to 0.
(3) Event misra_c_2012_rule_14_3_violation: | The condition "status == IOT_MQTT_SUCCESS" must be true.
```

File: libraries/standard/common/include/iot_linear_containers.h [hunk 4]
```diff
@@ -623,7 +623,8 @@ static inline IotLink_t * IotListDouble_FindFirstMatch( const IotListDouble_t *
 {
     /* The const must be cast away to match this function's return value. Nevertheless,
      * this function will respect the const-ness of pStartPoint. */
-    IotLink_t * pCurrent = ( IotLink_t * ) pStartPoint;
+    IotLink_t * pCurrent = ( IotLink_t * ) pStartPoint, * pMatchedLink = NULL;
```
Comment (gkwicker) [inline libraries/standard/common/include/iot_linear_containers.h]: Prefer these on separate lines.

File: libraries/standard/mqtt/src/iot_mqtt_subscription.c [hunk 5]
```diff
@@ -230,7 +230,7 @@ static bool _topicFilterMatch( const char * pTopicName,
                                const char * pTopicFilter,
                                uint16_t topicFilterLength )
 {
-    bool status = false;
+    bool status = false, matchFound = false;
```
Comment (gkwicker) [inline libraries/standard/mqtt/src/iot_mqtt_subscription.c]: Prefer on separate lines.

**Pull request merged**